drivers/lora/sx126x.c

```diff
@@ -0,0 +1,480 @@
+#include <drivers/lora.h>
+#include <drivers/spi.h>
+#include <zephyr.h>
+
+#include <sx126x/sx126x.h>
+
+#include "sx12xx_common.h"
+
+#define LOG_LEVEL CONFIG_LORA_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(sx126x);
+
+
+#if DT_HAS_COMPAT(semtech_sx1261)
+#define DT_DRV_COMPAT semtech_sx1261
+#define SX126X_DEVICE_ID SX1261
+#elif DT_HAS_COMPAT(semtech_sx1262)
+#define DT_DRV_COMPAT semtech_sx1262
+#define SX126X_DEVICE_ID SX1262
+#else
+#error No SX126x instance in device tree.
+#endif
+
+BUILD_ASSERT(DT_NUM_INST(semtech_sx1261) +
+	     DT_NUM_INST(semtech_sx1262) <= 1,
+	     "Multiple SX12xx instances in DT");
+
+#define GPIO_CS_LABEL		DT_INST_SPI_DEV_CS_GPIOS_LABEL(0)
+#define GPIO_CS_PIN		DT_INST_SPI_DEV_CS_GPIOS_PIN(0)
+
+#define GPIO_RESET_LABEL	DT_INST_GPIO_LABEL(0, reset_gpios)
+#define GPIO_RESET_PIN		DT_INST_GPIO_PIN(0, reset_gpios)
+#define GPIO_RESET_FLAGS	DT_INST_GPIO_FLAGS(0, reset_gpios)
+
+#define GPIO_BUSY_LABEL		DT_INST_GPIO_LABEL(0, busy_gpios)
+#define GPIO_BUSY_PIN		DT_INST_GPIO_PIN(0, busy_gpios)
+#define GPIO_BUSY_FLAGS		DT_INST_GPIO_FLAGS(0, busy_gpios)
+
+#define GPIO_IRQ_LABEL		DT_INST_GPIO_LABEL(0, irq_gpios)
+#define GPIO_IRQ_PIN		DT_INST_GPIO_PIN(0, irq_gpios)
+#define GPIO_IRQ_FLAGS		DT_INST_GPIO_FLAGS(0, irq_gpios)
+
+#define HAVE_GPIO_ANTENNA_ENABLE			\
+	DT_INST_NODE_HAS_PROP(0, antenna_enable_gpios)
+#define GPIO_ANTENNA_ENABLE_LABEL			\
+	DT_INST_GPIO_LABEL(0, antenna_enable_gpios)
+#define GPIO_ANTENNA_ENABLE_PIN				\
+	DT_INST_GPIO_PIN(0, antenna_enable_gpios)
+#define GPIO_ANTENNA_ENABLE_FLAGS			\
+	DT_INST_GPIO_FLAGS(0, antenna_enable_gpios)
+
+#if DT_INST_NODE_HAS_PROP(0, dio2_tx_enable)
+#define DIO2_TX_ENABLE DT_INST_PROP(0, dio2_tx_enable)
+#else
+#define DIO2_TX_ENABLE 0
+#endif
+
+BUILD_ASSERT(DIO2_TX_ENABLE,
+	     "Modules without DIO2 TX control currently unsupported");
+
+#define HAVE_DIO3_TCXO		DT_INST_NODE_HAS_PROP(0, dio3_tcxo_voltage)
+#if HAVE_DIO3_TCXO
+#define TCXO_DIO3_VOLTAGE	DT_INST_PROP(0, dio3_tcxo_voltage)
+#endif
+
+#if DT_INST_NODE_HAS_PROP(0, tcxo_power_startup_delay_ms)
+#define TCXO_POWER_STARTUP_DELAY_MS			\
+	DT_INST_PROP(0, tcxo_power_startup_delay_ms)
+#else
+#define TCXO_POWER_STARTUP_DELAY_MS	0
+#endif
+
+#define SX126X_CALIBRATION_ALL 0x7f
+
+struct sx126x_data {
+	struct device *reset;
+	struct device *busy;
+	struct device *irq;
+	struct gpio_callback irq_callback;
+	struct k_work irq_work;
+	DioIrqHandler *radio_dio_irq;
+#if HAVE_GPIO_ANTENNA_ENABLE
+	struct device *antenna_enable;
+#endif
+	struct device *spi;
+	struct spi_config spi_cfg;
+	RadioEvents_t radio_events;
+} dev_data;
+
+
+void SX126xWaitOnBusy(void);
+
+static int sx126x_spi_transceive(u8_t *req_tx, u8_t *req_rx, size_t req_len,
+				 void *data_tx, void *data_rx, size_t data_len)
+{
+	int ret;
+
+	const struct spi_buf tx_buf[] = {
+		{
+			.buf = req_tx,
+			.len = req_len,
+		},
+		{
+			.buf = data_tx,
+			.len = data_len
+		}
+	};
+
+	const struct spi_buf rx_buf[] = {
```

Comment:
You can use the same array of buffers for tx and rx. This would allow to drop `req_rx` argument. Maybe the same could be done with `data_rx`, because there is no case when `data_tx` and `data_rx` are passed simultaneously. In spi_nor.c and sx1276.c there is a `is_write` argument instead.

Reply:
I tried that strategy and it didn't seem to work on my nrf52 board. I suspect this is similar to an issues discussed on Slack when the sx1276 is used on that platform. I didn't debug it properly, but I think the DMA operation in the SPI peripheral gets confused when the same pointer is used in both the input and output register.